File: samples/cext-trey-research/src/model/apiModel.ts
import { Project, Consultant, Location } from './baseModel';

//#region GET requests for /projects --------------------

export interface ApiProjectAssignment {
    consultantName: string;
    consultantLocation: Location;
    role: string;
    forecastThisMonth: number;
    forecastNextMonth: number;
    deliveredLastMonth: number;
    deliveredThisMonth: number;
}

// Returned by all /api/projects GET requests
export interface ApiProject extends Project {
    consultants: ApiProjectAssignment[];
    forecastThisMonth: number;
    forecastNextMonth: number;
    deliveredLastMonth: number;
    deliveredThisMonth: number;
}
//#endregion

//#region GET requests for /me and /consultants ---

// Information about a project that a consultant is assigned to
export interface ApiConsultantAssignment {
    projectName: string;
    projectDescription: string;
    projectLocation: Location;
    clientName: string;
    clientContact: string;
    clientEmail: string;
    role: string;
    forecastThisMonth: number;
    forecastNextMonth: number;
    deliveredLastMonth: number;
    deliveredThisMonth: number;
}

// Returned by all /api/consultants GET requests
export interface ApiConsultant extends Consultant {
    projects: ApiConsultantAssignment[];
    forecastThisMonth: number;
    forecastNextMonth: number;
    deliveredLastMonth: number;
    deliveredThisMonth: number;
}
//#endregion

//#region POST request to /api/me/chargeTime ---
export interface ApiChargeTimeRequest {
    projectName: string;
    hours: number;
}
export interface ApiChargeTimeResponse {
    clientName: string;
    projectName: string;
    remainingForecast: number;
    message: string;
}
//#endregion

//#region POST request to /api/projects/assignConsultant ---
export interface ApiAddConsultantToProjectRequest {
    projectName: string;
    consultantName: string;
    role: string;
    hours: number;
}
export interface ApiAddConsultantToProjectResponse {
    clientName: string;
    projectName: string;
    consultantName: string;
    remainingForecast: number;
    message: string;
}
//#endregion

export interface ErrorResult {
    status: number;
    message: string;
}

File: samples/cext-trey-research-python/model/api_model.py
from dataclasses import dataclass, field
from typing import List
from model.base_model import Project, Consultant, Location

# GET requests for /projects

@dataclass
class ApiProjectAssignment:
    consultantName: str
    consultantLocation: Location
    role: str
    forecastThisMonth: float
    forecastNextMonth: float
    deliveredLastMonth: float
    deliveredThisMonth: float


@dataclass
class ApiProject(Project):
    consultants: List[ApiProjectAssignment] = field(default_factory=list)
    forecastThisMonth: float = 0.0
    forecastNextMonth: float = 0.0
    deliveredLastMonth: float = 0.0
    deliveredThisMonth: float = 0.0


# GET requests for /me and /consultants

@dataclass
class ApiConsultantAssignment:
    projectName: str
    projectDescription: str
    projectLocation: Location
    clientName: str
    clientContact: str
    clientEmail: str
    role: str
    forecastThisMonth: float
    forecastNextMonth: float
    deliveredLastMonth: float
    deliveredThisMonth: float


@dataclass
class ApiConsultant(Consultant):
    projects: List[ApiConsultantAssignment] = field(default_factory=list)
    forecastThisMonth: float = 0.0
    forecastNextMonth: float = 0.0
    deliveredLastMonth: float = 0.0
    deliveredThisMonth: float = 0.0


# POST request to /api/me/chargeTime

@dataclass
class ApiChargeTimeRequest:
    projectName: str
    hours: float


@dataclass
class ApiChargeTimeResponse:
    clientName: str
    projectName: str
    remainingForecast: float
    message: str


# POST request to /api/projects/assignConsultant

@dataclass
class ApiAddConsultantToProjectRequest:
    projectName: str
    consultantName: str
    role: str
    hours: float


@dataclass
class ApiAddConsultantToProjectResponse:
    clientName: str
    projectName: str
    consultantName: str
    remainingForecast: float
    message: str


@dataclass
class ErrorResult:
    status: int
    message: str
